catalog/harvest/catalog-harvest-api/src/main/java/org/codice/ddf/catalog/harvest/Harvester.java

```diff
@@ -0,0 +1,39 @@
+/**
+ * Copyright (c) Codice Foundation
+ *
+ * <p>This is free software: you can redistribute it and/or modify it under the terms of the GNU
+ * Lesser General Public License as published by the Free Software Foundation, either version 3 of
+ * the License, or any later version.
+ *
+ * <p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details. A copy of the GNU Lesser General Public
+ * License is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package org.codice.ddf.catalog.harvest;
+
+/**
+ * A {@code Harvester} is responsible for fetching remote resources and notifying any registered
+ * {@link Listener}s of create, update and delete events.
+ *
+ * <p>In the event that a {@code Harvester} goes down, it must remember the last state of the
+ * external server it was harvesting from. Upon a {@code Harvester} coming back up, it should only
```

Comment:
❓I’m not sure how strict we can be here. There’s always a chance that an `Harvester` ends up sending the same event twice to a its listeners no matter what, so we may want to change this to say that it is best effort and put a constraint on the listeners that they need to be idempotent. Thoughts?

Reply:
❓What happens when a listener registers the first time? Does it receive events for everything that already exists? Should that be an option the listeners provide when they register?
❓Does a harvester need to remember what events were sent to what listener? For instance if an event is sent to listener A but not B and the system goes down, is the expectation that only B will be sent the event upon restart?
❓How are listeners re-registered after a restart?

Reply:
❓To avoid the above problems, should we just allow a single listener and assume it’s always the same one that registers and unregisters?

Reply:
I like the idempotent idea.

Currently each harvester has its own listener. A listener would only receive events upon registration, no past events yet. I ended up having to register listeners before first harvest to avoid this right now. For the first pass I think we should allow single listener like you suggested and build on it later.

Reply:
❓Should we add a note in the comments about the fact that `Harvester` implementations may thread out calls to the listeners if needed and that listeners don't need to?